Windows: Path canonicalization in indexing — `kind` detection + memory_dir comparisons

## Background

Indexing-related tests fail on Windows where path comparisons across different canonicalization layers (drive letter case, short-name vs long-name, separator) don't match. Symptoms include \`kind=None\` where \`'memory'\` was expected, chunks not getting deleted because their source paths don't compare equal to \`memory_dirs\`, etc.

## Symptoms

\`\`\`
FAILED test_web_routes.py::TestSources::test_kind_set_when_source_under_memory_dir
  AssertionError: assert None == 'memory'

FAILED test_web_routes.py::TestRemoveMemoryDirChunkCleanup::test_delete_chunks_true_removes_matching_source_files
  assert 0 == 4

FAILED test_web_routes.py::TestUploadsUsage::test_populated
  assert 0 == 2
\`\`\`

The pattern: chunks have a source path stored at index time; later code compares against a configured \`memory_dir\`. On POSIX both go through \`Path.resolve()\` consistently. On Windows, the resolution may differ (drive case, \`\\\\?\\\\\` long-path prefix, short \`8.3\` form) so equality fails.

## Affected files (estimated)

- \`packages/memtomem/tests/test_indexing_engine.py\` — 11 failures
- \`packages/memtomem/tests/test_web_routes.py::TestSources\` / \`TestRemoveMemoryDirChunkCleanup\` / \`TestUploadsUsage\`
- Possibly: \`test_dirty.py\`, \`test_index_debounce.py\`

## Fix direction

1. Audit the production code that compares \`chunk.metadata.source_file\` against \`memory_dirs\` — confirm both go through identical canonicalization (\`Path.resolve()\` after a consistent normalization).
2. Consider storing canonical paths at index time so later comparison is just \`==\`, no resolve-on-read race.
3. Test fixtures may need to use \`tmp_path.resolve()\` consistently to avoid \`8.3\` vs long-path mismatches.

## Refs

- #634 — Windows + macOS CI matrix expansion (umbrella)
- PR #638 — matrix infrastructure (Windows currently informational)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

--- a/packages/memtomem/src/memtomem/indexing/engine.py
+++ b/packages/memtomem/src/memtomem/indexing/engine.py
@@ -5,6 +5,7 @@ from __future__ import annotations
 import asyncio
 import contextlib
 import logging
+import os
 import time
 from collections.abc import Callable, Iterable
 from datetime import datetime, timezone
@@ -123,14 +124,23 @@ def _dir_creation_time_iso(p: Path) -> str | None:
 def norm_dir_prefix(d: str | Path) -> str:
     """Return the directory path normalized for ``str.startswith`` matching.
 
-    Adds a trailing slash so a configured dir ``/foo`` does not falsely
-    claim files under ``/foo-bar/...``. Always runs through
-    :func:`~memtomem.storage.sqlite_helpers.norm_path` (which resolves
-    symlinks and applies Unicode NFC) so the prefix shape matches the
-    source-side normalisation regardless of whether the dir currently
+    Adds a trailing ``os.sep`` (platform-native separator) so a configured
+    dir does not falsely claim files under a sibling sharing the same
+    prefix (e.g. ``/foo`` should not match ``/foo-bar/...``). Always runs
+    through :func:`~memtomem.storage.sqlite_helpers.norm_path` (which
+    resolves symlinks and applies Unicode NFC) so the prefix shape matches
+    the source-side normalisation regardless of whether the dir currently
     exists on disk — the chunks table holds resolved paths, and a
     configured-but-missing dir would otherwise compare in raw ``/tmp``
     form against resolved ``/private/tmp`` source paths on macOS.
+
+    The trailing-separator step uses ``os.sep`` rather than a hardcoded
+    ``"/"`` so the prefix is consistent with ``norm_path``'s output on
+    Windows, where ``Path.resolve()`` returns backslash-separated strings
+    (``C:\\Users\\foo``) — a hardcoded ``"/"`` would yield a mixed-form
+    prefix that never matches a native source path under
+    ``startswith`` (#647). On POSIX, ``os.sep == "/"`` so behaviour is
+    unchanged.
 
     Used by both :func:`memory_dir_stats` (which buckets chunks per
     configured dir) and :func:`resolve_owning_memory_dir` (which goes
@@ -142,8 +152,8 @@ def norm_dir_prefix(d: str | Path) -> str:
 
     p = Path(d).expanduser()
     base = norm_path(p)
-    if not base.endswith("/"):
-        base += "/"
+    if not base.endswith(os.sep):
+        base += os.sep
     return base
 
 

--- a/packages/memtomem/src/memtomem/web/routes/system.py
+++ b/packages/memtomem/src/memtomem/web/routes/system.py
@@ -571,10 +571,14 @@ async def remove_memory_dir(
                 # the schema's ``ON DELETE CASCADE``.
                 deleted_chunks = 0
                 if delete_chunks:
+                    from memtomem.indexing.engine import norm_dir_prefix
+
                     rows = await storage.get_source_files_with_counts()
-                    prefix = resolved_norm
-                    if not prefix.endswith("/"):
-                        prefix = prefix + "/"
+                    # Use the canonical helper so the trailing-separator
+                    # rule (``os.sep``, native form on Windows) stays in
+                    # one place; matches the comparison done by
+                    # :func:`resolve_owning_memory_dir` (#647).
+                    prefix = norm_dir_prefix(resolved)
                     for row in rows:
                         source_path = row[0]
                         if norm_path(source_path).startswith(prefix):

--- a/packages/memtomem/tests/test_indexing_engine.py
+++ b/packages/memtomem/tests/test_indexing_engine.py
@@ -2572,6 +2572,47 @@ class TestMemoryDirStats:
 
 
 # ===========================================================================
+# 11b. norm_dir_prefix — trailing separator pin (#647)
+# ===========================================================================
+
+
+class TestNormDirPrefix:
+    """Pin: the trailing separator must come from ``os.sep`` rather than
+    a hardcoded ``"/"``.
+
+    Without this pin a regression where someone "tidies" ``os.sep`` back
+    to ``"/"`` would still pass on POSIX (because ``os.sep == "/"`` there)
+    and only break on the Windows CI leg (#647). Each parameter row
+    monkey-patches ``norm_path`` and ``os.sep`` so the test exercises both
+    platform shapes from a single Linux/macOS dev box.
+    """
+
+    @pytest.mark.parametrize(
+        "fake_norm_path,fake_sep,expected",
+        [
+            # POSIX shape — append "/" if missing.
+            ("/foo/bar", "/", "/foo/bar/"),
+            # Already-trailing POSIX — no double separator.
+            ("/foo/bar/", "/", "/foo/bar/"),
+            # Windows shape — append "\\" (this is what os.sep == "\\" buys).
+            ("C:\\Users\\foo", "\\", "C:\\Users\\foo\\"),
+            # Already-trailing Windows — no double separator.
+            ("C:\\Users\\foo\\", "\\", "C:\\Users\\foo\\"),
+        ],
+    )
+    def test_appends_os_sep_not_hardcoded_slash(
+        self, monkeypatch, fake_norm_path, fake_sep, expected
+    ):
+        import os as _os
+
+        from memtomem.indexing import engine
+
+        monkeypatch.setattr("memtomem.storage.sqlite_helpers.norm_path", lambda _p: fake_norm_path)
+        monkeypatch.setattr(_os, "sep", fake_sep)
+        assert engine.norm_dir_prefix("ignored-by-mock") == expected
+
+
+# ===========================================================================
 # 12. resolve_owning_memory_dir — source → owning dir lookup
 # ===========================================================================
 

